poseidon-sponge.md

```diff
@@ -0,0 +1,240 @@
+# Poseidon sponge construction
+
+- Version: 0.0.0
+- Maintainer: Jack Grigg
+
+## Introduction
+
+This document specifies the Poseidon sponge construction, a duplex sponge [[BDPV2011]]
+instantiated with the [Poseidon permutation]. It also specifies several instantiations of
+the sponge for use in various hashing domains.
+
+[BDPV2011]: https://keccak.team/files/SpongeDuplex.pdf
+[Poseidon permutation]: ./poseidon-permutation.md
+
+## Notation and conventions
+
+- `[a, b)` means the sequence of integers from `a` inclusive to `b` exclusive.
+- `[x; N]` means an array of length `N` where every entry is filled with `x`.
+- `y[i]` means the element of the array `y` at index `i`.
+- `len(y)` means the length of the array `y`. For example, `len([x; N]) == N`.
+
+## Duplex sponge
+
+The duplex sponge `PoseidonSponge` is built around `PoseidonPermutation`, a permutation of
+width `T`, which encapsulates the specific Poseidon parameters for the sponge. It has a
+parameter `RATE`, the number of input field elements processed by each invocation of the
+permutation, which is at most `T - 1`. The value `C = T - RATE` is called the capacity.
+
+> TODO: Describe how the permutation instance and the rate relate to each other and to the
+> desired security parameter `M`.
+
+`PoseidonSponge` has several internal properties:
+- An internal state of `T` field elements.
+- A mode variable that is either `absorbing` or `squeezing`.
+- A cache of field elements corresponding to the current mode.
+- A duplex padding function that is applied to input elements prior to incorporating them
+  into the sponge state.
```

Comment:
@rozbb I think the idea here is actually not to define a compliant padding function, so that you can use this as a normal sponge function without padding pretty much (for a fixed-length API where the padding wouldn't matter).

Reply:
Is that really the case? Why define `duplex_padding` at all if it's not going to be used anywhere? Surely padding is necessary at some layer of usage. I think the hard part is deciding which layer

Reply:
yeah that's one of my other comment, the zero padding could be removed from the spec if the absorption is defined differently. If you check the last section "Hash function with variable-length input" you can see the real padding in action (grep for "Domain-separate from fixed-length hash by terminating the message with a constant")

Reply:
This is partly my fault; per my [other comment](https://github.com/C2SP/C2SP/pull/3#discussion_r702188518) I misunderstood the padding condition in the duplex paper, assumed the Poseidon authors were specifying sponge-compatible padding, and extended it in a duplex-compatible fashion. My goal in this spec was to make clear that the Poseidon constructions were all built around a sponge, as the lack of consideration for that point is what led to the multiple mutually-incompatible implementations.

Now that it appears that sponges and duplexes are sadly also mutually-incompatible, how should we go about structuring a single clear primitive? I definitely want this to be usable for continuous inputs, in a transcript context.

Reply:
> I definitely want this to be usable for continuous inputs, in a transcript context.

Yeah I think that's where the problem can arise. The way we do it in Mina is that we make sure that the input is non-ambiguous before hashing it. So basically every input absorbed is fixed-length (or padded to be the same length all the time) and the sponge is used as a normal hash function. That's one way to do it, you can also tag each input with their length (that's what TupleHash does).

The other way to do it is to tackle that at the construction's layer. And that's what the duplex construction does.

I don't think it would be safe to use a hybrid construction for continuous inputs, as you can create collisions by mixing non-padded with padded absorptions.

Reply:
BTW an example of a collision I had in mind if you absorb 3 inputs: 1 variable-length, 1 fixed-length, and 1 variable-length. And variable-length inputs are padded at the end, fixed-length inputs are not padded:

* absorb [1, x], [y], [z] (this will absorb [1, x, 1, y, z, 1])
* absorb [], [x], [1, y, z] (this will absorb [1, x, 1, y, z, z])

Reply:
Maybe that's the wrong wording. I think it's more about defining an API that can be used to build both duplex and sponge functions. Some low-level APIs will work (imagine just defining a permutation), while some higher-level APIs won't (the case here).